corehq/apps/case_importer/tasks.py

```diff
@@ -11,38 +11,35 @@
 from .exceptions import ImporterError
 from .tracking.analytics import get_case_upload_files_total_bytes
 from .tracking.case_upload_tracker import CaseUpload
+from .tracking.task_status import make_task_status_success
 from .util import get_importer_error_message, exit_celery_with_error_message
 
 
 @task(serializer='pickle', queue='case_import_queue')
 def bulk_import_async(config, domain, excel_id):
     case_upload = CaseUpload.get(excel_id)
+    result_stored = False
     try:
         case_upload.check_file()
-    except ImporterError as e:
-        return exit_celery_with_error_message(bulk_import_async, get_importer_error_message(e))
-
-    try:
         with case_upload.get_spreadsheet() as spreadsheet:
             result = do_import(spreadsheet, config, domain, task=bulk_import_async,
                                record_form_callback=case_upload.record_form)
 
         _alert_on_result(result, domain)
-
-        # return compatible with soil
-        return {
-            'messages': result
-        }
+        # save the success result into the CaseUploadRecord
+        case_upload.store_task_result(make_task_status_success(result))
+        result_stored = True
     except ImporterError as e:
         return exit_celery_with_error_message(bulk_import_async, get_importer_error_message(e))
     finally:
-        store_task_result.delay(excel_id)
+        if not result_stored:
```

Comment:
This flag could be eliminated if the code were restructured like
```py
try:
    ... # get result
    case_upload.store_task_result(make_task_status_success(result))
    # should return something?
except ImporterError as e:
    store_failed_task_result.delay(excel_id)
    return exit_celery_with_error_message(bulk_import_async, get_importer_error_message(e))
except:
    store_failed_task_result.delay(excel_id)
    raise
```
Note that I renamed `store_task_result_if_failed` to `store_failed_task_result` in this snippet since we know it failed if it gets called. It may still check if there is a valid result before saving, but that seems unlikely, right?

Reply:
OK yeah this is a good suggestion too. I'll add to the followup PR

Reply:
Actually as I dug into it, the change in flow above also subtly change the order of things so that `exit_celery_with_error_message` is run _after_ firing off the `store_failed_task_result` task. I'll think a bit more about whether that matters, and not in the followup PR